### PR TITLE
[Hudi] Fix duplicate record naming in schema conversion

### PR DESCRIPTION
**Pull request by anniewang-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [x] Other (Hudi)

## Description

This PR fixes a bug in the Delta->Hudi schema conversion for structs. Previously, any table conversion that included at least one struct somewhere in the schema would fail. This is because the previous code was incorrectly naming the Avro RecordSchemas. There were two problems:

1.  There was no namespace added to each record, so for a struct column with the following schema:
     ` (myName STRUCT<myName: STRUCT<field1: INT>>)`
there would be an error due to the duplicate naming of the two structs. (**Avro does not allow recordSchemas with the same name under the same namespace**) It would incorrectly place the nested struct in the same namespace as the parent struct, and even though our schema should be valid the Avro schema creation would fail.

2. For some reason we were naming each record by its data type name instead of its own name. Since we represent the Delta schema as a struct, even if we don't have any nested structs inside our table, as long as we have at least one struct column in the schema we will end up creating a nested struct. Both of these records would be named "struct" and be under the same namespace (due to problem 1), so we would run into a duplication error even if we just have a single struct of ints.

So for an example table defined as follows:
`CREATE TABLE myTable (col1 STRUCT<field1: INT, field2: STRING>)`
the previous code would not work because it would represent our overall schema as a struct with name "struct", and our struct column would be a nested struct with name "struct" under the same namespace.

Now, I have changed it so that it works and is compatible with Spark+Hudi. We are now using namespaces and also naming with column names rather than column type names. For this example, our Avro schema would look like this:

```
{
  "type": "record",
  "name": "table",
  "fields": [
    {
      "name": "col1",
      "type": [
        "null",
        {
          "type": "record",
          "name": "col1",
          "namespace": "table",
          "fields": [
            {
              "name": "field1",
              "type": [
                "null",
                "int"
              ]
            },
            {
              "name": "field2",
              "type": [
                "null",
                "string"
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

## How was this patch tested?
Unit test and manually tested with Hudi SparkSession reader.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Comment by harperjiang** [comment]
May I have an example describing what the field names for struct look like before and after your changes?

File: hudi/src/main/scala/org/apache/spark/sql/delta/hudi/HudiSchemaUtils.scala [hunk 1]
```diff
@@ -33,16 +33,18 @@ object HudiSchemaUtils extends DeltaLogging {
...
-    def transform[E <: DataType](elem: E, isNullable: Boolean): Schema = elem match {
+    def transform[E <: DataType](elem: E, isNullable: Boolean, currentPath: String): Schema =
+      elem match {
       case StructType(fields) =>
 
         val avroFields: util.List[Schema.Field] = fields.map(f =>
           new Schema.Field(
             f.name,
-            transform(f.dataType, f.nullable),
+            transform(f.dataType, f.nullable,
```
Comment (harperjiang) [inline hudi/src/main/scala/org/apache/spark/sql/delta/hudi/HudiSchemaUtils.scala:43]: What is the impact of this change to the reader side?

**Review by harperjiang — COMMENTED** [review]
Please explain why this change has no impact to the external reader.

**Comment by anniewang-db** [comment]
> Please explain why this change has no impact to the external reader.

I guess technically it does have an impact because now converting a struct does work, whereas it previously did not. The expected user experience did not change though. I have changed the description to reflect this

File: hudi/src/main/scala/org/apache/spark/sql/delta/hudi/HudiSchemaUtils.scala (hunk 1 above)
Comment (anniewang-db) [reply hudi/src/main/scala/org/apache/spark/sql/delta/hudi/HudiSchemaUtils.scala]: No change other than the functionality works - their tables should look the same as if they created it in Hudi

**Comment by harperjiang** [comment]
> > Please explain why this change has no impact to the external reader.
> 
> I guess technically it does have an impact because now converting a struct does work, whereas it previously did not. The expected user experience did not change though. I have changed the description to reflect this

So what is the difference in your changed avro and the avro generated by using Hudi in Spark? Are they the same?

**Comment by anniewang-db** [comment]
> > > Please explain why this change has no impact to the external reader.
> > 
> > 
> > I guess technically it does have an impact because now converting a struct does work, whereas it previously did not. The expected user experience did not change though. I have changed the description to reflect this
> 
> So what is the difference in your changed avro and the avro generated by using Hudi in Spark? Are they the same?

Functionally the same - verified with Hudi reader

**Comment by anniewang-db** [comment]
jenkins retest this please

**Review by harperjiang — APPROVED** [review]
lgtm

**Pull request merged**